### PR TITLE
ARROW-3566: [Format] Clarify the type of dictonary encoded field

### PR DESCRIPTION
**Pull request by icexelloss**
Per discussion on https://github.com/apache/arrow/pull/2681#issuecomment-431337993

### CODE REVIEW

**Comment by icexelloss** [comment]
cc @wesm @elahrvivaz

**Comment by elahrvivaz** [comment]
@wesm I assume the C++ implementation encodes dictionary field types the same way, otherwise we wouldn't be passing the dictionary vector integration tests

File: format/Schema.fbs [hunk 1]
```diff
@@ -252,6 +252,7 @@ table Field {
   // Name is not required, in i.e. a List
   name: string;
   nullable: bool;
+  // This is the type of the index if the field is dictionary encoded
```
Comment (BryanCutler) [inline format/Schema.fbs]: I don't think this is right. In Java, when a dictionary Field is written it first calls `DictionaryUtility.toMessageFormat` [here](https://github.com/apache/arrow/blob/master/java/vector/src/main/java/org/apache/arrow/vector/util/DictionaryUtility.java#L45) which changes the Field to be the dictionary type.

Comment (elahrvivaz) [reply format/Schema.fbs]: Ah, thanks, that was the bridge to fix the gap between the java API and the IPC format. If I remember now, people didn't want to embed dictionary-related logic throughout the codebase, which would be required if the internal java field type has the dictionary type.

Comment (icexelloss) [reply format/Schema.fbs]: Ok I have checked the integration tests and it seems @BryanCutler is correct, the type here is the type of decoded type, not the index type:

This is taken from the generated json file in integration test 
```
      {
        "name": "dict1_0",
        "type": {
          "name": "utf8"
        },
        "nullable": true,
        "children": [],
        "dictionary": {
          "id": 0,
          "indexType": {
            "name": "int",
            "isSigned": true,
            "bitWidth": 8
          },
          "isOrdered": false
        }
      },
      ...
```

Comment (icexelloss) [reply format/Schema.fbs]: @wesm Could you please help confirm what the proper type should be here?

**Comment by icexelloss** [comment]
Ok I have checked the integration tests and it seems the type here is the type of decoded type, not the index type:

This is taken from the generated json file in integration test:

```
      {
        "name": "dict1_0",
        "type": {
          "name": "utf8"
        },
        "nullable": true,
        "children": [],
        "dictionary": {
          "id": 0,
          "indexType": {
            "name": "int",
            "isSigned": true,
            "bitWidth": 8
          },
          "isOrdered": false
        }
      },
      ...
```
@wesm Would you please help clarify what the correct type is here?

**Comment by wesm** [comment]
In C++ we have a synthetic `DictionaryType` that holds the dictionary / dictionary type as well as the index type. In the IPC metadata, the field type is that of the decoded type. The `DictionaryEncoding` table holds the index type

**Comment by icexelloss** [comment]
@wesm Thanks for the clarification. Then I think this PR should be good to go. 

I will need to go back to #2681 to see what to do there. Thanks all !

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**